### DISCUSSION TITLE
Function to compare hashes

### DISCUSSION THREAD

**Discussion by XMol** [puppetlabs/community#22] [Show and tell]
Puppet is incapable of comparing two hashes with each other and I'm very surprised that there is no function in the stdlib tool set to address that either. A function like this seems quite simple to me:
```ruby
Puppet::Functions.create_function(:cmp_hash) do
  dispatch :cmp_hashes_as_arrays do
    param 'Hash[Any, Any]', :hash
    param 'Hash[Any, Any]', :hash
  end

  dispatch :cmp_hashes_with_block do
    param 'Hash[Any, Any]', :hash
    param 'Hash[Any, Any]', :hash
    block_param 'Callable[1,1]', :block
  end

  def cmp_hashes_as_arrays (h1, h2)
    cmp_hashes_with_block(h1, h2) { |h| h.to_a.sort }
  end

  def cmp_hashes_with_block (h1, h2)
    yield(h1) <=> yield(h2)
  end
end
```

Given a lambda that transforms a Hash into something that Ruby can compare with `<=>`, arbitrary Hashes can be compared to each other. If no lambda is given, the functions falls back to casting both Hashes to sorted Arrays and then compare them (which Ruby can do already).

### Examples

```puppet
$hash1 = {
  'primary_key' => 2,
  'key1' => ['val1', 'val2'], 
  'key2' => { 'key3' =>  'val3', }, 
  'key4' => true,
  'key5' => 12345,
 }

$hash2 = {
  'primary_key' => 1,
  'key6' => ['val1', 'val2'], 
  'key7' => { 'key8' =>  'val9', }, 
  'key10' => true,
  'key11' => 67890,
 }

cmp_hash($hash1, $hash2) # => -1; 'key1' is the smallest key, thus $hash1 is smaller
cmp_hash($hash1, $hash2) |$h| { $h['primary_key'] } # => 1; $hash2 has the smaller value for 'primary_key', hence $hash1 is bigger
```
Now it is also possible to order an Array of Hashes with the built-in `sort()` function:
```puppet
$::disks.values.sort |$a, $b| { cmp_hash($a, $b) |$h| { $h['size_bytes'] } }
```

**Comment by binford2k** [comment]
I'm surprised you didn't [contribute your functions to stdlib](https://github.com/puppetlabs/puppetlabs-stdlib/pulls).

**Reply by XMol** [reply]
Because this seems such an obvious omission, I figured that there is a good reason for it that I'm not aware of. Or there simply is no demand for comparing Hashes in Puppet.

**Reply by b4ldr** [reply]
i dont think there is an equivalent in stdlib or puppet core so +1 for sending a PR

**Reply by XMol** [reply]
Thanks for the encouragement. 😊 Maybe I'll find the time for making a PR next week.

**Reply by XMol** [reply]
It took me longer than a week, but a [PR](https://github.com/puppetlabs/puppetlabs-stdlib/pull/1376) is now submitted.

**Comment by XMol** [comment]
Credits to @smortex, who proposed a solution that a accomplishes the comparison of two Hashes without adding a new function to Puppet:
```puppet
compare($hash1.dig('primary_key'), $hash2.dig('primary_key'))
```
